drivers/video/video_mcux_mipi_csi2rx.c

```diff
@@ -182,12 +163,136 @@ static int mipi_csi2rx_get_caps(const struct device *dev, enum video_endpoint_id
 	return video_get_caps(config->sensor_dev, ep, caps);
 }
 
+static inline int mipi_csi2rx_set_ctrl(const struct device *dev, unsigned int cid, void *value)
+{
+	const struct mipi_csi2rx_config *config = dev->config;
+
+	if (config->sensor_dev) {
+		return video_set_ctrl(config->sensor_dev, cid, value);
+	}
+
+	return -ENOTSUP;
+}
+
+static int mipi_csi2rx_set_frmival(const struct device *dev, enum video_endpoint_id ep,
+				   struct video_frmival *frmival)
+{
+	const struct mipi_csi2rx_config *config = dev->config;
+	int ret;
+
+	ret = video_set_frmival(config->sensor_dev, ep, frmival);
+	if (ret) {
+		LOG_ERR("Cannot set sensor_dev frmival");
+		return ret;
+	}
+
+	ret = mipi_csi2rx_update_settings(dev, ep);
+
+	return ret;
+}
+
+static int mipi_csi2rx_get_frmival(const struct device *dev, enum video_endpoint_id ep,
+				   struct video_frmival *frmival)
+{
+	const struct mipi_csi2rx_config *config = dev->config;
+
+	return video_get_frmival(config->sensor_dev, ep, frmival);
+}
+
+static uint64_t mipi_csi2rx_cal_frame_size(const struct video_format *fmt)
+{
+	return fmt->height * fmt->width * video_pix_fmt_bpp(fmt->pixelformat) * 8;
+}
+
+static uint64_t mipi_csi2rx_estimate_pixel_rate(const struct video_frmival *cur_fmival,
+						const struct video_frmival *fie_frmival,
+						const struct video_format *cur_format,
+						const struct video_format *fie_format,
+						uint64_t cur_pixel_rate, uint8_t laneNum)
+{
+	return mipi_csi2rx_cal_frame_size(cur_format) * fie_frmival->denominator *
+	       cur_fmival->numerator * cur_pixel_rate /
+	       (mipi_csi2rx_cal_frame_size(fie_format) * fie_frmival->numerator *
+		cur_fmival->denominator);
+}
+
+static int mipi_csi2rx_enum_frmival(const struct device *dev, enum video_endpoint_id ep,
+				    struct video_frmival_enum *fie)
+{
+	const struct mipi_csi2rx_config *config = dev->config;
+	struct mipi_csi2rx_data *drv_data = dev->data;
+	int ret;
+	uint64_t cur_pixel_rate, est_pixel_rate;
+	struct video_frmival cur_frmival;
+	struct video_format cur_fmt;
+
+	ret = video_enum_frmival(config->sensor_dev, ep, fie);
+	if (ret) {
+		return ret;
+	}
+
+	ret = video_get_ctrl(config->sensor_dev, VIDEO_CID_PIXEL_RATE, &cur_pixel_rate);
+	if (ret) {
+		LOG_ERR("Cannot get sensor_dev pixel rate");
+		return ret;
+	}
```

Comment:
Maybe there is an inconsistency bewteen the `video_get_ctrl()` API and the `_enum_frmival()` API as used here?

It does math on the [*current* pixel rate](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/video/ov5640.c#L376-L380) and the *queried* format from `fie` (which is not applied yet).

Would that allow a pixel rate for a wrong format to be used in the frame interval math?
Should the pixel rate become part of the `fie` query struct so that it can be filled by `sensor_dev`?
Should the `fie` not contain any format and the current format always be used to simplify this situation?

It is very possible that I missed something too...

Reply:
- the current pixel rate is coming from the sensor and is only updated whenever there is a change in frame rate, so when enum_frmival is called, it always uses the latest / updated value.

> - Should the fie not contain any format and the current format always be used to simplify this situation?

=> No, fie is used to query supported framerates for _any_ given format, not only the current one. Thats why it must contain the format in the query.